qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,849 @@
+    The time ``t`` and ``duration`` are in units of dt, i.e. system cycle time,
+    and this function is sampled with a discrete time vector in [0, ``duration``]
+    sampling the pulse envelope at every 0.5 dt (middle sampling strategy) when
+    :meth:`get_waveform` method is called.
+    The function can generate complex-valued samples representing a phase in Gaussian plane,
+    and real and imaginary part may be supplied to two quadratures of
+    the IQ mixer in the control electronics.
+    The sample data is not generated until the get method is called thus
+    a symbolic pulse instance only stores parameter values and waveform shape,
+    which greatly reduces memory footprint during the program generation.
+
+
+    .. rubric:: Constraint functions
+
+    This is defined with a class method :meth:`_define_constraints` that returns
+    a list of symbolic expressions. These constraint functions are called
+    when a symbolic pulse instance is created with assigned parameter values with ``limit``
+    representing the policy of allowing the maximum amplitude exceeding 1.0
+    (defaults to ``True``, i.e. it doesn't allow amplitude > 1.0).
+    The symbolic pulse subclass will be instantiated
+    only when all constraint functions return ``True``.
+    This means the return value of these symbolic expressions must be boolean.
+    Any number of expression can be defined in the class method.
+    When branching logic (if clause) is necessary, one can use `SymPy ITE`_ class
+    or a list of expressions consisting of expression for "IF", "THEN", and "ELSE".
+    Typically, it evaluates the maximum pulse amplitude so that it doesn't exceed the
+    signal dynamic range [-1.0, 1.0]. When validation is not necessary,
+    this method can return an empty list.
+
+    .. _SymPy ITE: https://docs.sympy.org/latest/modules/logic.html#sympy.logic.boolalg.ITE
+
+
+    .. rubric:: Examples
+
+    This is an example of how a pulse author can define new pulse subclass.
+
+    .. code-block:: python
+
+        import sympy as sym
+
+        class CustomPulse(SymbolicPulse):
+
+            PARAM_DEF = ["duration", "p0", "p1"]
+
+            def __init__(self, duration, p0, p1, name):
+                super().__init__(duration=duration, parameters=(p0, p1), name=name)
+
+            @classmethod
+            def __define_envelope(cls):
+                duration, t, p0, p1 = sym.symbols("duration, t, p0, p1")
+                ...
+
+                return ...
+
+            @classmethod
+            def _define_constraints(cls):
+                limit, duration, p0, p1 = sym.symbols("limit, duration, p0, p1")
+
+                const1 = ...
+                const2 = ...
+
+                return [const1, const2, ...]
+
+    Then a user can create the ``CustomPulse`` instance as follows.
+
+    .. code-block:: python
+
+        my_pulse = CustomPulse(duration=160, p0=1.0, p1=2.0)
```

Comment:
Can you add an example with Parameters?

Reply:
Simply you can directly pass parameters to the p0 and p1, which is totally same as current `ParametricPulse`. I'll avoid writing parameter section since there is no mechanism to assign parameters to pulse instance (that has been deprecated). Now parameter assignment can be done through schedule. Since this is API doc of pulse instance, handling of parameters should be explained in other pulse tutorials.